### ISSUE TITLE
Feature Request: Full File/Folder Management (DAM)

### ISSUE THREAD

**Issue by batonac** [darktable-org/darktable#12907]
The biggest single feature I continuously long for in Darktable is the ability to manage folder hierarchies directly within the app. This is a standard feature of many other photo workflow applications, including Lightroom, Digikam, ACDSee, (the late) Picasa, etc.

I would summarize the feature request this way. Give Darktable the ability to:
1. Import entire folder structures (folders and subfolders) and filter library by these hierarchies.
2. Create and remove folders, move (and remove) files around between them.
3. Refresh the contents of the folders to update the photo collection with the latest files. (either manually or automatically*)
4. Track the file system location of a file by retaining collection assignments even as it's moved on the file system.

*ACDSee promotes "no import" in their suite, "Unlike the Adobe® products, ACDSee Photo Studio does not require you to laboriously import photo collections. Simply install ACDSee on your hard-drive and point it to your photos – no juggling or running batches – it’s just install and create awesomeness." ... that would be really cool in Darktable as well.

Proof of concept:
Lightroom:
![image](https://user-images.githubusercontent.com/4996285/201917432-b7eecd19-fa54-4fff-860b-ca857b46f119.png)
Digikam:
![image](https://user-images.githubusercontent.com/4996285/201917542-c7954bc5-e85d-41cd-8667-854e794e9d53.png)
Picasa:
![image](https://user-images.githubusercontent.com/4996285/201917298-4564eab1-5a43-4b93-99d2-491af70a72a4.png)

**Comment by difrkaguilar** [comment]
This is precisely one of the functions of some software that I like the least. I don't want a program in which I would dedicate myself to developing to do file manager functions. For example in a photo session, in many occasions I previously select in camera with stars the possible pictures to develop, then, I have to import all the images and then delete the ones I don't want (unfortunately darktable doesn't have the function to visualize the camera ratings from the import window) and keep only a few, the ones selected in camera and some others.

At no time I would be in favor of a program that makes a scan by folders if I am not going to develop all the images, that would be a process that the program is going to do for pleasure because not all the images that I have in my organized folders are going to be developed. Picasa is one of the programs that I hate the most. it makes a search of all and each one of the images that there is in the system, it takes hours indexing all the disks depending on the hardware and amount of folders that the user has, I don't see it feasible at all.

darktable is not a file manager, for that each operating system has one expressly created for that purpose and I would not want to delete images that I might need later. When working on a project basis as most photographers do, there is no need to create a file manager just for this purpose.

More than these functions I am attracted by the possibility of Capture One 23 that has just been released where from the import process itself you can make a selection by checkmarks, colors and star ratings and it is also possible to view the photos at 100% size, all this before importing without the need to import them all to a catalog or to the discs to start editing.

Anyway, I am not a developer, and they are the most qualified to define this feature request, since they would be the ones who would do it in any case, although I would not like darktable to take that way.

Sorry for my English, I'm using DeepL.com

**Comment by ralfbrown** [comment]
(1) already exists.  You can import an entire directory tree by checking "recursive", and you can filter by an entire directory tree either by holding down the appropriate modifier key while clicking on the top-level directory in the folder view in the Collections module, or manually adding "/%" to the folder name in the text box.

(3) already exists in the manual version.  Simply re-import a folder or folder tree and only the new images are added.  There is also a script to remove unavailable images from the database.

(4)  would be extremely difficult to implement unless you are willing to dramatically slow down **all** imports. How are you to tell that a random file on the disk is the same as one with a completely different name in a completely different folder that was previously added to the database?  The only way I can think of is to store a SHA hash for each image, and that will require reading the entire image to compute the hash, rather than just a bit of metadata.

**Comment by batonac** [comment]
>At no time I would be in favor of a program that makes a scan by folders if I am not going to develop all the images, that would be a process that the program is going to do for pleasure because not all the images that I have in my organized folders are going to be developed. Picasa is one of the programs that I hate the most. it makes a search of all and each one of the images that there is in the system, it takes hours indexing all the disks depending on the hardware and amount of folders that the user has, I don't see it feasible at all.

@difrkaguilar I can understand why this strikes you as undesirable. It certainly would be unfortunate if new features were added that became an encumbrance to your workflow. I probably shouldn't have brought Picasa into it, as it's not a pro-level tool. I was simply looking around for some UI-examples.

I would simply submit to you that there are other workflows where this could be a big asset, and it need not get in the way of anyone's workflow if implemented well.

I'm not suggesting that the importer works any differently, but rather that the filesystem hierarchy is available as an additional filter tool once the import is complete.

>darktable is not a file manager, for that each operating system has one expressly created for that purpose and I would not want to delete images by mistake that I might need later. When working on a project basis as most photographers do, there is no need to create a file manager just for this purpose.

IMO, the major issue with the file managers from every major operating system vendor, is that none of them do very well at metadata handling, including filtering and updating metadata on files. This becomes a big issue for many photography workflows.

In addition, there is not file manager, that I am aware of, which allows for optionally viewing and managing assets in the current folder and sub-folders simultaneously. Lightroom does this beautifully, and this allows creative workers to quickly filter down the assets for a potential selection at the level they want, using the file system as the filter.

This is where a good photo-management app comes to the rescue.

>(1) already exists. You can import an entire directory tree by checking "recursive", and you can filter by an entire directory tree either by holding down the appropriate modifier key while clicking on the top-level directory in the folder view in the Collections module, or manually adding "/%" to the folder name in the text box.

@ralfbrown yeah, the importer, including recursive functions, is great. As stated above, my biggest issue is with the filtering capabilities once the photos are imported. I can't find anywhere where I can see the hierarchy, as such. Am I doing something wrong?

>(3) already exists in the manual version. Simply re-import a folder or folder tree and only the new images are added. There is also a script to remove unavailable images from the database.

That's fantastic, but a "resync folder" context menu option would still be an improvement IMO.

>(4) would be extremely difficult to implement unless you are willing to dramatically slow down all imports. How are you to tell that a random file on the disk is the same as one with a completely different name in a completely different folder that was previously added to the database? The only way I can think of is to store a SHA hash for each image, and that will require reading the entire image to compute the hash, rather than just a bit of metadata.

I understand what you're saying, and I should have clarified. Point 4 should probably have been a sub-point under point 2. I wasn't referring so much to files moved outside of Darktable, but rather moved through Darktable.

---

I believe the strongest case for this functionality is for grouping files together (by client, project, product, etc), especially in a collaborative workspace. The filesystem layer acts as the most basic-level structure for photo collection(s), giving creative workers the ability to contribute to the overall organization of the assets, regardless of how they are accessed and managed, and allows for quickly importing and contributing back to these structures without porting the database from one workstation to another.

Hopefully that makes sense. I want to stress that this need not interfere with, or alter, existing workflows in Darktable. Simply adding the ability to filter by hierarchies and manage those folder structures would expand the reach of this great program.

**Comment by ralfbrown** [comment]
> @ralfbrown yeah, the importer, including recursive functions, is great. As stated above, my biggest issue is with the filtering capabilities once the photos are imported. I can't find anywhere where I can see the hierarchy, as such. Am I doing something wrong?

Have you switched the collections module from "film roll" as criterion to "folder"?  Once in folder mode, you'll see the imported hierarchy, and hovering the mouse on the folder name in the top pulldown (not the hierarchy display) will give a tooltip with instructions for how to include/exclude subdirectories.

**Comment by deric** [comment]
I think @batonac has some valid points, maybe we don't have to call it Full File Manager a "Folder Viewer" might be a good start. 

Having "the only correct way for accessing file on disk" could be very frustrating for new users. And users don't  know how to use the ["collections" module](https://news.ycombinator.com/item?id=31961371):

> I believe I somehow imported my entire /pictures folder at some point, producing many thousands of images in a film roll; I just downloaded 4.0.0 and have spent some time trying to figure out how to reset the film roll, or darktable altogether. Or how to create a new library.

There might be multiple modules on the left side of Lighttable the would allow browsing photos on disk(s). Each user might prefer different workflow, different way of storing their photos. This doesn't have to be a breaking change for existing users.

The collections module in "folder" mode works fine (though a bit hard to find), but there few things that might be improved. E.g. expanding the view vertically might challenging for a new user as there is no arrow when hovering over the bottom border.

![collections](https://user-images.githubusercontent.com/138573/203770542-70ad2452-54bc-4bf7-b0e1-923ae598ae26.png)

I guess for people coming from Lightroom there might be confusing the need to "import" photos into a single database file called catalog (which contains all modifications and might get corrupted). This isn't (hopefully) coming to Darktable.
It might be important for new users to understand the non-destructive edits and sidecar files. 

Currently navigating in deep tree structures in Darktable is hard. It would be nice to have a tree view of folders where the top level would your photos root (without full path to that folder). Also creating subfolders would be nice. Currently right-click on a folder collapses the structure before showing the context menu, which is (at least for me) unwanted behavior.

**Comment by webghostx** [comment]
In my opinion, the Collections module is not implemented good enough to replace an image management system. It would be so nice to work with a directory tree without constantly falling back on rolls of film and then having to reorient and move around every time. And I would very much welcome an option for auto import within a defined directory.

For over a year, I've been trying to find a comfortable workflow to manage images. I wanted to switch to Digikam for this, but unfortunately that's only good for exported images, since Darktable's edits can't be displayed.

My workflow:
After a day of taking photos, I first move all the pictures from the memory card to my folder for the recordings. So I create a folder like Pictures/Recordings/23.02.23_div. I then import the folder into Darktable. The next thing I want to do is sort the images, and it's getting tedious. I mark some pictures, go to "move" and create another folder like Pictures/Recordings/23.02.23_something.

The problems:
What happens now is that you drop out of the directory view back into the film reel view. If you switch back, you have to shimmy through the directory tree again. The same problem if you completely empty a directory, i.e. delete all images. Then you are thrown out and have to reorient yourself again. Another problem: if you are in a higher-level directory and delete a couple images from it, it can happen that all rolls of film below are removed. Then you have to import everything again, which can take hours

My conclusion: Darktable is great for editing images, but very tedious to manage them

**Comment by webghostx** [comment]
@Nilvus can you tell me where to find the collection module in the code?

**Comment by Nilvus** [comment]
Should be src/libs/collect.c file.

Note that there's some questions to rework completely collection module. Nothing is sure and that part is quite sensible and could quickly become controversial, as you can see on that issue (that's why I closed that. If you want to work on, I suggest you to search all discussions about that part here, and on pixls.us. And especially those links:

#10643
https://discuss.pixls.us/t/collection-filters-proposals/32995
https://discuss.pixls.us/t/discussion-around-collection-filters-improvements/32489

**Comment by Nilvus** [comment]
Oh, and taht PR: #10733

**Comment by emko** [comment]
why must we be force to use these collection thing? why can't we be allowed to make folders and move images around? at the very least allow a context menu on right click on a image to open folder in system file browser... and one to go to this folder in lighttable for cases when your using a filter in collections and want to see the other photos in the folder it came from. I am trying digikam for this but then it does not work well with dt 

why can't we have both file based and collection based management?

**Comment by wpferguson** [comment]
> why must we be force to use these collection thing?

Two of the first things built in darktable were film rolls and collections, approximately 14 years ago.  The rest of darktable was built on that foundation.

To change it would be at best extremely difficult and at worst impossible.  But, you're welcome to try :smiley: 

> why can't we be allowed to make folders and move images around?

You can select images in lighttable, then use the `move...` button in the selected images module to put them in another directory.  There are also lua scripts to move images (rename_images.lua and transfer_hiearchy.lua).

> at the very least allow a context menu on right click on a image to open folder in system file browser.

There is a lua script to do this, OpenInExplorer.lua.  You could assign a shortcut to it.

> why can't we have both file based and collection based management?

Because someone has to develop it.  I'm sure that some developers have looked at it before and were either unwilling to take on a project that large, or unable to figure out a way to do it well and satisfy all parties.

**Comment by emko** [comment]
> > why must we be force to use these collection thing?
> 
> Two of the first things built in darktable were film rolls and collections, approximately 14 years ago. The rest of darktable was built on that foundation.
> 
> To change it would be at best extremely difficult and at worst impossible. But, you're welcome to try smiley
> 
> > why can't we be allowed to make folders and move images around?
> 
> You can select images in lighttable, then use the `move...` button in the selected images module to put them in another directory. There are also lua scripts to move images (rename_images.lua and transfer_hiearchy.lua).
> 
> > at the very least allow a context menu on right click on a image to open folder in system file browser.
> 
> There is a lua script to do this, OpenInExplorer.lua. You could assign a shortcut to it.
> 
> > why can't we have both file based and collection based management?
> 
> Because someone has to develop it. I'm sure that some developers have looked at it before and were either unwilling to take on a project that large, or unable to figure out a way to do it well and satisfy all parties.

ahhh my bad, i thought it was just dt not wanting to do it at all, i see now the issue is that no dev has implemented it yet, thanks for the suggestions. I really did give collections system a try for around 6 months and it just was not for me
